backends/p4tools/modules/testgen/targets/bmv2/map_direct_externs.cpp

```diff
@@ -1,46 +1,68 @@
-    // Try to either get counters or meters from the table.
-    const auto *impl = table->properties->getProperty("meters");
-    if (impl == nullptr) {
-        impl = table->properties->getProperty("counters");
+std::optional<const IR::Declaration_Instance *> MapDirectExterns::getExternFromTableImplementation(
+    const IR::Property *tableImplementation) {
+    const auto *selectorExpr = tableImplementation->value->to<IR::ExpressionValue>();
+    if (selectorExpr == nullptr) {
+        ::error("Extern property is not an expression %1%", tableImplementation->value);
+        return std::nullopt;
     }
-    if (impl == nullptr) {
-        return false;
-    }
-    // Cannot map a temporary mapped direct extern without a counter.
-    const auto *selectorExpr = impl->value->checkedTo<IR::ExpressionValue>();
+    // If the extern expression is a constructor call it is not relevant.
     if (selectorExpr->expression->is<IR::ConstructorCallExpression>()) {
-        return true;
+        return std::nullopt;
     }
     // Try to find the extern in the declared instances.
     const auto *implPath = selectorExpr->expression->to<IR::PathExpression>();
-    auto implementationLabel = implPath->path->name.name;
-
+    if (implPath == nullptr) {
+        ::error("Invalid extern path %1%", selectorExpr->expression);
+        return std::nullopt;
+    }
     // If the extern is not in the list of declared instances, move on.
-    auto it = declaredExterns.find(implementationLabel);
+    auto it = declaredExterns.find(implPath->path->name.name);
     if (it == declaredExterns.end()) {
-        return true;
+        ::error("Cannot find direct extern declaration %1%", implPath);
+        return std::nullopt;
     }
     // BMv2 does not support direct externs attached to multiple tables.
-    const auto *declInstance = it->second;
-    if (directExternMap.find(declInstance) != directExternMap.end()) {
-        FATAL_ERROR(
-            "Direct extern was already mapped to a table. It can not be attached to two tables. ");
-        return true;
+    auto mappedTable = directExternMap.find(it->second->controlPlaneName());
+    if (mappedTable != directExternMap.end()) {
+        ::error(
+            "Direct extern %1% was already mapped to table %2%. It can not be "
+            "attached to two tables.",
+            it->second, mappedTable->second);
+        return std::nullopt;
     }
-    directExternMap.emplace(declInstance, table);
-    return true;
+    return it->second;
+}
+
+bool MapDirectExterns::preorder(const IR::P4Table *table) {
+    // Try to get extern implementation properties from the table.
+    for (const auto *tableProperty : kTableExternProperties) {
+        const auto *impl = table->properties->getProperty(tableProperty);
+        if (impl != nullptr) {
+            auto declInstanceOpt = getExternFromTableImplementation(impl);
+            if (!declInstanceOpt.has_value()) {
+                return false;
```

Comment:
This is a bit strange to me. I would've thought you'd just want to continue so that you can test all extern properties? Or, if having an implementation defined, yet not having that extern not be defined is super illegal, then maybe better to allow this function to fail?

This seems like a strange limbo state.

Reply:
Good catch, that was not intentional.

Reply:
There are indeed some cases where not returning a declaration instance has a benign cause where we do not throw an error. Iteration should not stop there.